contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -3209,6 +3236,295 @@ pltsql_store_func_default_positions(ObjectAddress address, List *parameters, con
+		
+		/* search for the row for update => build the key */
+		ScanKeyInit(&key[0],
+					Anum_bbf_schema_perms_dbid,
+					BTEqualStrategyNumber, F_INT2EQ,
+					Int16GetDatum(dbid));
+		ScanKeyEntryInitialize(&key[1], 0,
+					Anum_bbf_schema_perms_schema_name,
+					BTEqualStrategyNumber, InvalidOid,
+					tsql_get_server_collation_oid_internal(false),
+					F_TEXTEQ, CStringGetTextDatum(logical_schema_name));
+		ScanKeyEntryInitialize(&key[2], 0,
+					Anum_bbf_schema_perms_object_name,
+					BTEqualStrategyNumber, InvalidOid,
+					tsql_get_server_collation_oid_internal(false),
+					F_TEXTEQ, CStringGetTextDatum(object_name));
+		ScanKeyEntryInitialize(&key[3], 0,
+					Anum_bbf_schema_perms_object_type,
+					BTEqualStrategyNumber,
+					InvalidOid,
+					tsql_get_server_collation_oid_internal(false),
+					F_TEXTEQ,
+					CStringGetTextDatum(object_type));
+
+		/* scan */
+		scan = systable_beginscan(bbf_schema_rel,
+				get_bbf_schema_perms_oid(),
+				false, NULL, 4, key);
+
+		/* get the scan result -> original tuple */
+		tuple_bbf_schema = systable_getnext(scan);
+
+		/* Build updated parameter list */ 
+		func_args = gen_func_arg_list_from_params(parameters);
+
+		/*
+		* If a permission on the same object is granted to multiple grantees,
+		* there can be multiple rows in the catalog corresponding to each grantee name.
+		* All such rows need to be updated with the new parameters.
+		*
+		* It is OK to not throw an error if an entry is not found in 'babelfish_schema_permissions'.
+		* Explaination: An entry is added to 'babelfish_schema_permissions' only if an object has an explicit GRANT on it.
+		* It is not necessary that each RENAME on an object has a GRANT of that object too.
+		* Hence, there can be missing entries.
+		*/
+		while (HeapTupleIsValid(tuple_bbf_schema))
+		{
+			/* create new tuple to substitute */
+			new_record_bbf_schema[Anum_bbf_schema_perms_function_args - 1] = CStringGetTextDatum(func_args);
+			new_record_repl_bbf_schema[Anum_bbf_schema_perms_function_args - 1] = true;
+
+			new_tuple = heap_modify_tuple(tuple_bbf_schema,
+										bbf_schema_dsc,
+										new_record_bbf_schema,
+										new_record_nulls_bbf_schema,
+										new_record_repl_bbf_schema);
+
+			CatalogTupleUpdate(bbf_schema_rel, &new_tuple->t_self, new_tuple);
+
+			heap_freetuple(new_tuple);
+			tuple_bbf_schema = systable_getnext(scan);
+		}
+		if (physical_schema_name != NULL)
+			pfree(physical_schema_name);
+		if (logical_schema_name != NULL)
+			pfree(logical_schema_name);
+		if (object_name != NULL)
+			pfree(object_name);
+
+		systable_endscan(scan);
+	}
+	table_close(bbf_schema_rel, RowExclusiveLock);
+}
+
+/*
+ * Update 'funcsignature', 'default_positions', 'flag_validity', 'flag_values', 
+ * 'modify_date', and 'definition' in 'babelfish_function_ext'
+ */
+void
+alter_bbf_function_ext(ObjectAddress address, ObjectWithArgs *owa, List *parameters, const char *queryString, int origname_location)
```

Comment:
Same as above, can we instead modify the original function that handles babelfish_function_ext to deal with replacement instead of creating a new function that copies much of the same code?

Reply:
Just removed this function. Using pltsql_store_func_default_positions for both cases now